### ISSUE TITLE
I want to add UART support

### ISSUE THREAD

**Issue by r2aiv** [svofski/bk0010#17]
Hello! I want to add custom UART support to this project. How i can do it?

**Comment by svofski** [comment]
Hi!
I think you can start by checking out how it's done in the simulation for test printout:
https://github.com/svofski/bk0010/blob/master/vm1/icarus/simme.v

Low-level implementations of UART can be found elsewhere. In vector06cc I used txd by Tobias Gubener:
https://github.com/svofski/vector06cc/blob/master/src/floppy/uart/txd.vhd

I'm not familiar with specifics of the ИРПС controller, which I'm assuming you want to implement, so I can't guide you here but I'm happy to help where I can.

**Comment by r2aiv** [comment]
Ok. Thank you for reply. Can we go to email and communicate on Russian? =)

**Comment by svofski** [comment]
Можно продолжать по-русски прямо тут.

**Comment by r2aiv** [comment]
Вот и прекрасно, я просто не знаю правил гитхаба, вдруг этот тут возбраняется? ) Мысль такая - мне просто надо как-то "сесть" на шину МПИ, чтобы добавлять свои модули, начать хочу с UART'а. Кстати, в БКшке присутвует порт ТЛГ - готовый программный (?) UART. В этом проекте, насколько понимаю, он не реализован?

**Comment by svofski** [comment]
Шина МПИ никак особенно не реализована. Если задача прикрутить что-то в рамках проекта, то все достаточные сигналы есть и так.

Вот выборка регистров по адресу:
https://github.com/svofski/bk0010/blob/master/src/bkcore.v#L210

Если хочется совместимость с внешним миром, чтобы например настоящий модуль воткнуть, тут придется попотеть, чтобы все сигналы сделать совместимыми. Сейчас для этого вообще ничего нет, совместимость с реальным железом я не планировал.

ТЛГ это наверное и есть ИРПС. Логически тот же UART, электрически отличается тем, что это токовая петля (MIDI фактически, только скорости отличаются). Он присуствует в БК-0010Ш, а в бытовых вариантах по-моему не встречается. Не реализован.

Я думаю за описанием устройств на логическом уровне можно отправиться к эмуляторам. BKBTL например, я уверен, реализует сильно более широкий спектр устройств, чем мой скромный проект:
https://github.com/nzeemin/bkbtl-doc
https://github.com/nzeemin/bkbtl

**Comment by r2aiv** [comment]
Нет, ТЛГ и ИРПС - разные вещи. ИРПС это ВП1-035 с обвязкой, в виде внешнего блока, а ТЛГ выведен (не везде) на ноги порта УП. За инфу спасибо! Исходники изучу )

**Comment by svofski** [comment]
Я только вот это нашел про ТЛГ:
http://bk0010.org/forum/?id=1999&old
И так и не понял пока, что это такое. Помню, что еще с настоящей БК меня этот вопрос беспокоил.

**Comment by r2aiv** [comment]
По сути, две точки в схеме (RX и TX), где их брать именно - уже не помню, распаиваются на разъем УП и получается программный UART, управляемый через специально отведенное программное прерывание EMT. И, кстати, да - kt315b на том форуме - это я. Я правильно понимаю, что в проекте на DE1 есть возможность загрузки с линейного входа?

**Comment by r2aiv** [comment]
А не подскажете, как реализован перехват EMT36. Можно ли малой кровью его с магнитофона загрузить?

**Comment by svofski** [comment]
Конечно. Вот отсюда и дальше собствернно перехват:

https://github.com/svofski/bk0010/blob/6b7cbb1a6805b23340fe2a86eaa3e1f843c45f8a/src/bkcore.v#L260

Перехват переключает процессор в "режим ядра". Это лично моя хака, она документирована вот тут:

https://github.com/svofski/bk0010/wiki/Memory_Management

(хм, таблицы немного пострадали при переезде с googlecode). 
Это сделано по мотивам настоящего MMU от PDP-11, но только по мотивам.  Через нее же реализована начальная загрузка и вообще все общение с SD-картой.

Ломать не строить впрочем. Чтобы сделать загрузку с кассетофона надо просто оторвать перехват emt36 и оставить таким образом стандартный обработчик из юзермоды. Провод от звукового кодека в соответствующий порт, насколько я могу видеть, уже протянут: tape_in в bkcore.v. Но я не могу вспомнить, работала такая загрузка, или нет.

**Comment by r2aiv** [comment]
Спасибо! То, что надо ) При подключении магнитофона к линейному входу аудиокодека соответствующие биты регистра 177716 начинают "дергаться" в соответствии с записью - пока все, что могу сказать. На выход тоже работает (звук то идет - куда ему еще деться?). Вобщем, буду пробовать, спасибо за инфу. О результатах сообщу )

**Comment by svofski** [comment]
В vector06cc вход подключен аналогичным образом и там загрузка с магнитофона работает. Собственно сюда звук 1 в 1 перенесен с Вектора. Но в там у меня процессор точно повторяет такты, как в оригинале, а тут сделана только аппроксимация производительности. Это может пагубно сказаться на процедурах замера интервала, задержки и тому подобного.

**Comment by r2aiv** [comment]
А ведь самое забавное, что заработало ) И пишет и читает. Правда с магнитофоном хороших результатов добиться не удалось - но это издержки его и  самой кассеты. Но на комп пишет и читает отлично. Так же грузит на ура wav'ы, сгенерированные эмулятором Калмыкова, хотя они отличаются по скорости.
